Invalid signature in class fixture causes difficult-to-debug error

### minimal reproduction

```python
import pytest


class TestCase:
    @pytest.fixture(autouse=True)
    def fix():  # XXX: should have `self`
        pass
```

### output (python2)

```
$ pytest t.py
============================= test session starts ==============================
platform linux2 -- Python 2.7.15rc1, pytest-3.8.1, py-1.6.0, pluggy-0.7.1
rootdir: /tmp, inifile:
collected 0 items / 1 errors                                                   

==================================== ERRORS ====================================
____________________________ ERROR collecting t.py _____________________________
venv/local/lib/python2.7/site-packages/funcsigs/__init__.py:71: in signature
    raise ValueError('invalid method signature')
E   ValueError: invalid method signature
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!
=========================== 1 error in 0.11 seconds ============================
```

### output (python3)

```
$ pytest t.py 
============================= test session starts ==============================
platform linux -- Python 3.6.5, pytest-3.8.1, py-1.6.0, pluggy-0.7.1
rootdir: /tmp, inifile:
collected 0 items / 1 errors                                                   

==================================== ERRORS ====================================
____________________________ ERROR collecting t.py _____________________________
/usr/lib/python3.6/inspect.py:3037: in signature
    return Signature.from_callable(obj, follow_wrapped=follow_wrapped)
/usr/lib/python3.6/inspect.py:2787: in from_callable
    follow_wrapper_chains=follow_wrapped)
/usr/lib/python3.6/inspect.py:2198: in _signature_from_callable
    return _signature_bound_method(sig)
/usr/lib/python3.6/inspect.py:1788: in _signature_bound_method
    raise ValueError('invalid method signature')
E   ValueError: invalid method signature
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!
=========================== 1 error in 0.30 seconds ============================
```

```
$ pip freeze --all
atomicwrites==1.2.1
attrs==18.2.0
funcsigs==1.0.2
future-fstrings==0.4.4
more-itertools==4.3.0
pathlib2==2.3.2
pip==18.0
pluggy==0.7.1
py==1.6.0
pytest==3.8.1
scandir==1.9.0
setuptools==40.4.3
six==1.11.0
tokenize-rt==2.0.1
wheel==0.31.1
```

--- a/changelog/4026.bugfix.rst
+++ b/changelog/4026.bugfix.rst
@@ -1,0 +1,1 @@
+Improve error message when it is not possible to determine a function's signature.

--- a/src/_pytest/compat.py
+++ b/src/_pytest/compat.py
@@ -13,7 +13,7 @@ from contextlib import contextmanager
 import py
 
 import _pytest
-from _pytest.outcomes import TEST_OUTCOME
+from _pytest.outcomes import TEST_OUTCOME, fail
 from six import text_type
 import six
 
@@ -131,9 +131,17 @@ def getfuncargnames(function, is_method=False, cls=None):
     # ordered mapping of parameter names to Parameter instances.  This
     # creates a tuple of the names of the parameters that don't have
     # defaults.
+    try:
+        parameters = signature(function).parameters
+    except (ValueError, TypeError) as e:
+        fail(
+            "Could not determine arguments of {!r}: {}".format(function, e),
+            pytrace=False,
+        )
+
     arg_names = tuple(
         p.name
-        for p in signature(function).parameters.values()
+        for p in parameters.values()
         if (
             p.kind is Parameter.POSITIONAL_OR_KEYWORD
             or p.kind is Parameter.KEYWORD_ONLY

--- a/testing/test_collection.py
+++ b/testing/test_collection.py
@@ -952,3 +952,23 @@ def test_collect_init_tests(testdir):
             "*<Function 'test_foo'>",
         ]
     )
+
+
+def test_collect_invalid_signature_message(testdir):
+    """Check that we issue a proper message when we can't determine the signature of a test
+    function (#4026).
+    """
+    testdir.makepyfile(
+        """
+        import pytest
+
+        class TestCase:
+            @pytest.fixture
+            def fix():
+                pass
+    """
+    )
+    result = testdir.runpytest()
+    result.stdout.fnmatch_lines(
+        ["Could not determine arguments of *.fix *: invalid method signature"]
+    )

